deepdive1/README.md

```diff
@@ -0,0 +1,371 @@
+
+## Why are unit tests important?
+Say you wrote 23 units tests in your code. You then add a new feature, and now suddenly 5 of your unit tests aren't passing anymore. Unit tests are a clear indicator to show how solidly built your code is, and are very helpful for debugging. You can quickly catch errors in your code and lets you verify that the logic is correct. They might seem superfluous, but they're easy to implement and prevent you from backtracking/spending a lot of time trying to figure out where things went wrong.
+
+Testing is something you'll most likely encounter when you get a software job, so learning how to do testing now will give you a leg up in the game.
+
+## What are some things I want to test in my code?
+Pretty much everything! You can test both backend and frontend components. Generally you'd write a test everytime you write a piece of code that does something. Made a new fragment that should show you 10 pictures? You can write a test for that.
+
+## What are some rules I should follow for writing tests?
+You want your tests to be:
+* **Targeted** - You test one thing at a time
+* **Isolated** - The code you're testing should be isolated from the rest of the app's code and any external dependencies or events
+* **Repeatable and predictable** - The result of your test should be predictable and the test should be repeatable
+* **Independent** - Your tests can run in any order and shouldn't be dependent on other tests
+
+## What are the different overall types of tests?
+* **Unit tests** target small, isolated pieces of code, like functions. A unit test might check that a function with a return statement returns the correct value, or that a function which modifies a list does so correctly.
+* **Integration tests** combine isolated units of code to ensure they work together correctly. An integration test might save a record to a database, then read it back and check that it's unchanged.
+* **End-to-end tests** (also known as **system tests**) simulate user interactions with the app and check that it's working as it should. An end-to-end test might fill out a login form, click "Sign In", and verify that the correct activity has loaded afterwards.
+
+A common philosophy (check out [this blog post](https://testing.googleblog.com/2015/04/just-say-no-to-more-end-to-end-tests.html) from Google) suggests a "testing pyramid": a large base of unit tests, some integration tests, and fewer end-to-end tests. The goal of the "testing pyramid" is to catch a bug with the smallest test possible. For example, an end-to-end test might fail if a function it depends on isn't returning the right value, but it won't give you much information about where the failure is; a unit test for the function will point you towards the failure right away.
+
+(Thanks OlinJS for that wonderfully written section :heart:)
+
+## [Test-Driven Development (TDD)](https://www.agilealliance.org/glossary/tdd/)
+TDD is an Agile practice where you:
```

Comment:
Is it weird to throw in the word "Agile" without explaining what it is?  It's not really important that they know what Agile is, but it might throw them off to see an unknown word.